include/seqan3/argument_parser/detail/version_check.hpp

```diff
@@ -0,0 +1,415 @@
+                    return;
+                }
+            }
+        }
+
+        std::array<int, 3> empty_version{0, 0, 0};
+        std::array<int, 3> srv_app_version{};
+        std::array<int, 3> srv_seqan_version{};
+
+        std::ifstream version_file{cookie_path / (name + ".version")};
+
+        if (version_file.is_open())
+        {
+            std::string line{};
+            std::getline(version_file, line); // get first line which should only contain the version number of the app
+
+        if (line != UNREGISTERED_APP)
+            srv_app_version = get_numbers_from_version_string(line);
+#if !defined(NDEBUG) || defined(SEQAN3_TEST_VERSION_CHECK_)
+        else
+            std::cerr << MESSAGE_UNREGISTERED_APP;
+#endif // !defined(NDEBUG) || defined(SEQAN3_TEST_VERSION_CHECK_)
+
+            std::getline(version_file, line); // get second line which should only contain the version number of seqan
+            srv_seqan_version = get_numbers_from_version_string(line);
+
+            version_file.close();
+        }
+
+#if !defined(NDEBUG) || defined(SEQAN3_TEST_VERSION_CHECK_) // only check seqan version in debug or testing mode
+        if (srv_seqan_version != empty_version)
+        {
+            std::array<int, 3> seqan_version = {SEQAN3_VERSION_MAJOR, SEQAN3_VERSION_MINOR, SEQAN3_VERSION_PATCH};
+
+            if (seqan_version < srv_seqan_version)
+                std::cerr << MESSAGE_SEQAN3_UPDATE;
+        }
+#endif
+
+        if (srv_app_version != empty_version) // app version
+        {
+#if defined(NDEBUG) || defined(SEQAN3_TEST_VERSION_CHECK_) // only check app version in release or testing mode
+            if (get_numbers_from_version_string(version) < srv_app_version)
+                std::cerr << MESSAGE_APP_UPDATE;
+#endif // defined(NDEBUG) || defined(SEQAN3_TEST_VERSION_CHECK_)
+
+#if !defined(NDEBUG) || defined(SEQAN3_TEST_VERSION_CHECK_) // only notify developer that app version should be updated on server
+            if (get_numbers_from_version_string(version) > srv_app_version)
+                std::cerr << MESSAGE_REGISTERED_APP_UPDATE;
+#endif // !defined(NDEBUG) || defined(SEQAN3_TEST_VERSION_CHECK_)
+        }
+
+        std::cerr << std::flush;
+
+        if (program.empty())
+        {
+            prom.set_value(false);
+            return;
+        }
+
+        // launch a separate thread to not defer runtime.
+        if (cookie_line == "ALWAYS")
+            std::thread(call_server, "", command, std::move(prom)).detach();
+        else
+            std::thread(call_server, timestamp_filename, command, std::move(prom)).detach();
+
+    }
+
+    //!\brief The identification string that may appear in the version file if an app is unregistered.
+    static constexpr char const * const UNREGISTERED_APP = "UNREGISTERED_APP";
+    //!\brief The message directed to the developer of the app if a new seqan3 version is available.
+    static constexpr char const * const MESSAGE_SEQAN3_UPDATE =
```

Comment:
why not use string_views?

Reply:
What's the advantage to char const *?

Reply:
It's a c++ type, you can call .size() on it, you can pass it around by reference and not worry about. You make less mistakes when placing the `const` ... :laughing: